test-app/tests/integration/components/headless-form-native-validation-test.gts

```diff
@@ -0,0 +1,754 @@
+/* eslint-disable simple-import-sort/imports,padding-line-between-statements,decorator-position/decorator-position -- Can't fix these manually, without --fix working in .gts */
+
+import { blur, click, fillIn, render } from '@ember/test-helpers';
+import { module, test } from 'qunit';
+
+import HeadlessForm from 'ember-headless-form/components/headless-form';
+import sinon from 'sinon';
+import { setupRenderingTest } from 'test-app/tests/helpers';
+
+import type { RenderingTestContext } from '@ember/test-helpers';
+
+module(
+  'Integration Component HeadlessForm > Native validation',
+  function (hooks) {
+    setupRenderingTest(hooks);
+
+    interface TestFormData {
+      firstName?: string;
+      lastName?: string;
+    }
+
+    test('form has novalidate to render custom validation errors', async function (assert) {
+      const data: TestFormData = {};
+      const submitHandler = sinon.spy();
+
+      await render(<template>
+        <HeadlessForm @data={{data}} @onSubmit={{submitHandler}} as |form|>
+          <form.field @name="firstName" as |field|>
+            <field.label>First Name</field.label>
+            <field.input required data-test-first-name />
+          </form.field>
+          <button type="submit" data-test-submit>Submit</button>
+        </HeadlessForm>
+      </template>);
+
+      assert.dom('form').hasAttribute('novalidate');
+    });
+
+    test('onSubmit is not called when validation fails', async function (assert) {
+      const data: TestFormData = {};
+      const submitHandler = sinon.spy();
+
+      await render(<template>
+        <HeadlessForm @data={{data}} @onSubmit={{submitHandler}} as |form|>
+          <form.field @name="firstName" as |field|>
+            <field.label>First Name</field.label>
+            <field.input required data-test-first-name />
+          </form.field>
+          <button type="submit" data-test-submit>Submit</button>
+        </HeadlessForm>
+      </template>);
+
+      await click('[data-test-submit]');
+
+      assert.false(
+        submitHandler.called,
+        '@onSubmit is not called when required field in empty'
+      );
+    });
+
+    test('onInvalid is called when validation fails', async function (assert) {
+      const data: TestFormData = {};
+      const invalidHandler = sinon.spy();
+
+      await render(<template>
+        <HeadlessForm @data={{data}} @onInvalid={{invalidHandler}} as |form|>
+          <form.field @name="firstName" as |field|>
+            <field.label>First Name</field.label>
+            <field.input required data-test-first-name />
+          </form.field>
+          <button type="submit" data-test-submit>Submit</button>
+        </HeadlessForm>
+      </template>);
+
+      await click('[data-test-submit]');
+
+      assert.true(
+        invalidHandler.calledOnce,
+        '@onInvalid was called when required field in empty'
```

Comment:
what arguments does it receive? should those be asserted here as well?

Reply:
Good catch, yeah, I was a bit sloppy here 😅

This receives the same arguments as for custom validations like in [this test](https://github.com/CrowdStrike/ember-headless-form/blob/main/test-app/tests/integration/components/headless-form-validation-test.gts#L196-L207), but the thing is as mentioned elsewhere that the actual validation message is brittle, so I'd test for the _existence_ of it, but not the actual value. But I was not sure how to do this easily with sinon. Any suggestion?

Reply:
if it's tested elsewhere, that should be fine 💯